yatima-rs/src/expression.rs

```diff
@@ -169,6 +170,53 @@ impl Expr {
     env.insert_expr(cid, self)?;
     Ok(cid)
   }
+
+  pub fn shift(self, inc: &Nat, dep: &Option<Nat>) -> Self {
+    if *inc == Nat::zero() {
+        return self.clone()
+    }
+    match self {
+      Self::Var(nam, idx) => match dep {
+        // only increment free variables
+        Some(dep) if idx < *dep => Self::Var(nam, idx),
+        _ => Self::Var(nam, idx + inc),
+      },
+      Self::Lam(nam, bind, typ, bod) => {
+        Self::Lam(
+          nam,
+          bind,
+          Box::new((*typ).shift(inc, dep)),
+          // must increment depth because we are within the scope of the new binder `nam`
+          Box::new((*bod).shift(inc, &dep.as_ref().map(|x| x + Nat::one())))
+        )
+      }
+      Self::App(fun, arg) => {
+        Self::App(Box::new(fun.shift(inc, dep)), Box::new(arg.shift(inc, dep)))
+      }
+      Self::Pi(nam, bind, dom, rng) => {
+        Self::Pi(
+          nam,
+          bind,
+          Box::new(dom.shift(inc, dep)), Box::new(rng.shift(inc, &dep.as_ref().map(|x| x + Nat::one()))),
+        )
+      }
+      Self::Let(nam, typ, exp, bod) => {
+        Self::Let(
+          nam,
+          Box::new(typ.shift(inc, dep)),
+          Box::new(exp.shift(inc, dep)),
+          Box::new(bod.shift(inc, &dep.as_ref().map(|x| x + Nat::one()))),
+        )
+      }
+      Self::Fix(nam, x) => {
```

Comment:
(note: I'm not yet familiar with the use of this syntax, so not sure if this is correct)

Reply:
Oops, for some reason I assumed I already did it this way and that this was a confirmation 😆

See #26